### DISCUSSION TITLE
PytzUsageWarning > pytz migration

### DISCUSSION THREAD

**Discussion by ElissonRodrigues** [agronholm/apscheduler#570] [Q&A]
2021-10-29T23:13:15.908903+00:00 app[web.1]: /app/.heroku/python/lib/python3.10/site-packages/apscheduler/util.py:95: PytzUsageWarning: The zone attribute is specific to pytz's interface; please migrate to a new time zone provider. For more details on how to do so, see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
2021-10-29T23:13:15.908905+00:00 app[web.1]:   if obj.zone == 'local':
2021-10-29T23:13:15.909318+00:00 app[web.1]: /app/.heroku/python/lib/python3.10/site-packages/apscheduler/triggers/interval.py:66: PytzUsageWarning0: The normalize method is no longer necessary, as this time zone supports the fold attribute (PEP 495). For more details on migrating to a PEP 495-compliant implementation, see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
2021-10-29T23:13:15.909319+00:00 app[web.1]:   return self.timezone.normalize(next_fire_time)

**Comment by agronholm** [comment]
This happens with tzlocal 4.0. You can either (safely) ignore those warnings or downgrade tzlocal to 2.x. The problem is "properly" fixed on master but likely won't be in the 3.x branch.

**Reply by ElissonRodrigues** [reply]
OK, thank you.

**Reply by rtoma** [reply]
Happens with tzlocal 4.1 too

**Reply by LukeLIN-web** [reply]
Happens with tzlocal 4.2 too

**Comment by filipre** [comment]
I believe this issue causes the redis jobstore to fail. Could it be related?

```
/usr/local/lib/python3.9/site-packages/apscheduler/triggers/interval.py:66: PytzUsageWarning: The normalize method is no longer necessary, as this time zone supports the fold attribute (PEP 495). For more details on migrating to a PEP 495-compliant implementation, see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
  return self.timezone.normalize(next_fire_time)
Traceback (most recent call last):
  File "/app/renebot.py", line 94, in <module>
    renebot.start()
  File "/app/signalbot/bot.py", line 112, in start
    self.scheduler.start()
  File "/usr/local/lib/python3.9/site-packages/apscheduler/schedulers/asyncio.py", line 45, in start
    super(AsyncIOScheduler, self).start(paused)
  File "/usr/local/lib/python3.9/site-packages/apscheduler/schedulers/base.py", line 167, in start
    self._real_add_job(job, jobstore_alias, replace_existing)
  File "/usr/local/lib/python3.9/site-packages/apscheduler/schedulers/base.py", line 872, in _real_add_job
    store.add_job(job)
  File "/usr/local/lib/python3.9/site-packages/apscheduler/jobstores/redis.py", line 81, in add_job
    pipe.hset(self.jobs_key, job.id, pickle.dumps(job.__getstate__(),
TypeError: cannot pickle 'TaskStepMethWrapper' object
```

**Reply by agronholm** [reply]
The job store isn't failing. You're trying to add a job which contains an unpickleable argument. The warning above has nothing whatsoever to do with that.

**Reply by agronholm** [reply]
Btw, are you perhaps trying to schedule a coroutine object?

**Reply by filipre** [reply]
> Btw, are you perhaps trying to schedule a coroutine object?

Thank you! That was the issue.

**Comment by jonilsonds9** [comment]
Hi guys, has there been a fix for this problem yet?

**Reply by agronholm** [reply]
Is silencing the warnings not good enough as a workaround?

**Reply by jonilsonds9** [reply]
yes, it's enough, but I only asked if there is already a definitive solution, because I didn't find anything saying that it was fixed definitively

**Reply by agronholm** [reply]
It's fixed in master, but that's for v4.0.

**Reply by ghost** [reply]
> Is silencing the warnings not good enough as a workaround?

What's the recommended way to silence this warning?

**Reply by agronholm** [reply]
[Here](https://docs.python.org/3.10/library/warnings.html#overriding-the-default-filter) are some examples.

**Comment by agronholm** [comment]
This has become such a big noise generating problem that I decided to do something about it in the 3.x series. Those warnings are no longer present in the test suite. I will make a release shortly but feel free to validate the result by installing directly from GitHub.

**Comment by consonantal** [comment]
Hey guys, there is any improvement about applying to solving the official 3.x level @agronholm ?

**Reply by agronholm** [reply]
I released v3.9.0 yesterday which should fix this.

**Reply by agronholm** [reply]
..or at least tried to. The release workflow was not triggered because APScheduler uses a different tagging scheme than the rest of my projects (something I should fix).

**Reply by agronholm** [reply]
Ok, it's up now.

**Reply by consonantal** [reply]
Everything is ok with v3.9.1 now for me. Thanks for the solution @agronholm .

**Comment by Felix83000** [comment]
Hi, I tried 3.9.1, still the same Warning: 
<img width="1589" alt="Capture d’écran 2022-03-03 à 18 11 27" src="https://user-images.githubusercontent.com/16541780/156615654-95c3f320-5b16-441a-a27c-58b13316b49e.png">
<img width="667" alt="Capture d’écran 2022-03-03 à 18 12 06" src="https://user-images.githubusercontent.com/16541780/156615736-499020dd-c3b5-461e-afec-839d9b896a2b.png">

**Reply by agronholm** [reply]
Weird. I ran the test suite with `-W error`, with the latest tzlocal installed, no errors or warnings. Then I ran simple scripts using both interval and cron triggers and did not encounter this warning. Can you provide a script that would reproduce this warning?

**Reply by helios555** [reply]
I have the same issue. 
This script reproduces the warning for me, with 3.9.1 and tzlocal 4.2 and 4.1.

```python
from apscheduler.schedulers.background import BackgroundScheduler
from apscheduler.triggers.cron import CronTrigger
from importlib.metadata import version

print(f"apscheduler: {version('apscheduler')}")
print(f"tzlocal: {version('tzlocal')}") # tested with 4.1, 4.2

def job():
    print("Job execution")
scheduler = BackgroundScheduler()

# This doesnt throw warning
scheduler.add_job(job, CronTrigger.from_crontab('1-59 * * * *'))
scheduler.add_job(job, CronTrigger.from_crontab('* * * * *'))
# This does
scheduler.add_job(job, CronTrigger.from_crontab('1/3 * * * *'))
#scheduler.add_job(job, CronTrigger.from_crontab('1 * * * *'))
#scheduler.add_job(job, CronTrigger.from_crontab('1/5 * * * *'))
scheduler.start()
```

Output:
```
apscheduler: 3.9.1
tzlocal: 4.2
This doesnt throw warning 
This does
/home/helios/.local/lib/python3.8/site-packages/apscheduler/util.py:436: PytzUsageWarning: 
The localize method is no longer necessary, as this time zone supports the fold attribute (PEP 495). 
For more details on migrating to a PEP 495-compliant implementation, 
see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
  return tzinfo.localize(dt)
```

**Reply by Ailitonia** [reply]
> Weird. I ran the test suite with `-W error`, with the latest tzlocal installed, no errors or warnings. Then I ran simple scripts using both interval and cron triggers and did not encounter this warning. Can you provide a script that would reproduce this warning?

I have the same issue with 3.9.1 and tzlocal 4.1.
But it seems to only appeared in cron jobs.

This is my test script:
```python
import sys
import asyncio
from datetime import datetime
from importlib.metadata import version
from apscheduler.schedulers.asyncio import AsyncIOScheduler


scheduler: AsyncIOScheduler = AsyncIOScheduler()


async def test_cron_job() -> None:
    print(f'cron sleeping started at {datetime.now()}')
    await asyncio.sleep(1)
    print('cron sleeping end')


async def test_interval_job() -> None:
    print(f'interval sleeping started at {datetime.now()}')
    await asyncio.sleep(1)
    print('interval sleeping end')


async def main():
    scheduler.add_job(
        test_cron_job,
        'cron',
        second='*/2',
        id='test_cron_job'
    )

    scheduler.add_job(
        test_interval_job,
        'interval',
        seconds=2,
        id='test_interval_job'
    )

    scheduler.start()
    scheduler.print_jobs()
    await asyncio.sleep(1e10)


if __name__ == '__main__':
    print('system version:', sys.version)
    print('apscheduler version:', version('apscheduler'))
    print('tzlocal version:', version('tzlocal'))
    asyncio.run(main())
```

Output:
```
system version: 3.10.4 (tags/v3.10.4:9d38120, Mar 23 2022, 23:13:41) [MSC v.1929 64 bit (AMD64)]
apscheduler version: 3.9.1
tzlocal version: 4.1
Jobstore default:
    test_cron_job (trigger: cron[second='*/2'], next run at: 2022-05-03 18:56:24 CST)
    test_interval_job (trigger: interval[0:00:02], next run at: 2022-05-03 18:56:24 CST)
C:\my_project_path\venv\lib\site-packages\apscheduler\util.py:436: PytzUsageWarning: The localize method is no longer necessary, as this time zone supports the fold attribute (PEP 495). For more details on migrating to a PEP 495-compliant implementation, see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
  return tzinfo.localize(dt)
cron sleeping started at 2022-05-03 18:56:24.000732
interval sleeping started at 2022-05-03 18:56:24.299684
cron sleeping end
interval sleeping end
cron sleeping started at 2022-05-03 18:56:26.000694
interval sleeping started at 2022-05-03 18:56:26.295731
cron sleeping end
interval sleeping end
```

Only add cron job:
```python
import sys
import asyncio
from datetime import datetime
from importlib.metadata import version
from apscheduler.schedulers.asyncio import AsyncIOScheduler


scheduler: AsyncIOScheduler = AsyncIOScheduler()


async def test_cron_job() -> None:
    print(f'cron sleeping started at {datetime.now()}')
    await asyncio.sleep(1)
    print('cron sleeping end')


async def main():
    scheduler.add_job(
        test_cron_job,
        'cron',
        second='*/2',
        id='test_cron_job'
    )

    scheduler.start()
    scheduler.print_jobs()
    await asyncio.sleep(1e10)


if __name__ == '__main__':
    print('system version:', sys.version)
    print('apscheduler version:', version('apscheduler'))
    print('tzlocal version:', version('tzlocal'))
    asyncio.run(main())
```

Output:
```
system version: 3.10.4 (tags/v3.10.4:9d38120, Mar 23 2022, 23:13:41) [MSC v.1929 64 bit (AMD64)]
apscheduler version: 3.9.1
tzlocal version: 4.1
Jobstore default:
    test_cron_job (trigger: cron[second='*/2'], next run at: 2022-05-03 18:58:50 CST)
C:\my_project_path\venv\lib\site-packages\apscheduler\util.py:436: PytzUsageWarning: The localize method is no longer necessary, as this time zone supports the fold attribute (PEP 495). For more details on migrating to a PEP 495-compliant implementation, see https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html
  return tzinfo.localize(dt)
cron sleeping started at 2022-05-03 18:58:50.004531
cron sleeping end
cron sleeping started at 2022-05-03 18:58:52.015928
cron sleeping end
```

Only add interval job:
```python
import sys
import asyncio
from datetime import datetime
from importlib.metadata import version
from apscheduler.schedulers.asyncio import AsyncIOScheduler


scheduler: AsyncIOScheduler = AsyncIOScheduler()


async def test_interval_job() -> None:
    print(f'interval sleeping started at {datetime.now()}')
    await asyncio.sleep(1)
    print('interval sleeping end')


async def main():
    scheduler.add_job(
        test_interval_job,
        'interval',
        seconds=2,
        id='test_interval_job'
    )

    scheduler.start()
    scheduler.print_jobs()
    await asyncio.sleep(1e10)


if __name__ == '__main__':
    print('system version:', sys.version)
    print('apscheduler version:', version('apscheduler'))
    print('tzlocal version:', version('tzlocal'))
    asyncio.run(main())
```

Output:
```
system version: 3.10.4 (tags/v3.10.4:9d38120, Mar 23 2022, 23:13:41) [MSC v.1929 64 bit (AMD64)]
apscheduler version: 3.9.1
tzlocal version: 4.1
Jobstore default:
    test_interval_job (trigger: interval[0:00:02], next run at: 2022-05-03 18:59:59 CST)
interval sleeping started at 2022-05-03 18:59:59.578265
interval sleeping end
interval sleeping started at 2022-05-03 19:00:01.574353
interval sleeping end
```

**Reply by usherreuben** [reply]
@Ailitonia @Felix83000
The problem is pytz is considered deprecated in favor of zoneinfo module and its backports.
the Timezone format  on your system is in conflict when coverted to used by AsyncIOScheduler

**FIX IT WITH THIS**

import tzlocal

scheduler = AsyncIOScheduler( _timezone=str(tzlocal.get_localzone())_ )

**Reply by zhu0629** [reply]
did not work.
```
system version: 3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64 bit (AMD64)]
apscheduler version: 3.9.1.post1
tzlocal version: 4.2
```

**Reply by Declarcall** [reply]
os: ubuntu 20.04 lts, python version: 3.8, apscheduler version: 3.10.1 . the  same, not working.

> @Ailitonia @Felix83000 The problem is pytz is considered deprecated in favor of zoneinfo module and its backports. the Timezone format on your system is in conflict when coverted to used by AsyncIOScheduler
> 
> **FIX IT WITH THIS**
> 
> import tzlocal
> 
> scheduler = AsyncIOScheduler( _timezone=str(tzlocal.get_localzone())_ )

**Reply by agronholm** [reply]
Is there a reason why you can't just mute that specific warning?

**Comment by Felix83000** [comment]
I cannot provide a script because I am using it in Django.
With the script of @[Ailitonia](https://github.com/Ailitonia) could you gess and fix the issue?

Thanks

**Comment by bransonngai** [comment]
Same here, and apsheduler seems fail to run..

**Reply by agronholm** [reply]
Not helpful. It seems to be running for others.

**Comment by twisteroidambassador** [comment]
This warning is still present with apscheduler 3.10.0 and tzlocal 4.2

code location: site-packages/apscheduler/utils.py:428

**Reply by agronholm** [reply]
Did you not read my initial response?

**Reply by twisteroidambassador** [reply]
> Did you not read my initial response?

Yes, but I also read above that this should have been fixed in 3.9.0.

I noticed that the code location for the warning (utils.py:428) is different from the original post. Perhaps there is more than one place where the warning is triggered, and not all of them are fixed in version 3.

**Reply by agronholm** [reply]
Yeah, so this is not fully fixable without major refactoring, which I won't be doing in the 3.x series. Adding ignore settings for that warning is an acceptable solution on the 3.x series.

**Comment by shaonis** [comment]
This works for me - no warnings (might be useful for those who use UTC)

import pytz
from apscheduler.schedulers.asyncio import AsyncIOScheduler

scheduler = AsyncIOScheduler(timezone=pytz.utc)
